File: src/core/srs.ts
// SRS kick tables with tetr.io-style 180 kicks ("SRS+" 180 spins enabled).
// Offsets are (dx, dy) with y up, tried in order.

import type { PieceType, Rot } from './pieces';
import { cellsAt } from './pieces';
import type { Board } from './board';

type Kick = readonly [number, number];

// Key: `${from}>${to}`
const JLSTZ_KICKS: Record<string, Kick[]> = {
  '0>1': [[0, 0], [-1, 0], [-1, 1], [0, -2], [-1, -2]],
  '1>0': [[0, 0], [1, 0], [1, -1], [0, 2], [1, 2]],
  '1>2': [[0, 0], [1, 0], [1, -1], [0, 2], [1, 2]],
  '2>1': [[0, 0], [-1, 0], [-1, 1], [0, -2], [-1, -2]],
  '2>3': [[0, 0], [1, 0], [1, 1], [0, -2], [1, -2]],
  '3>2': [[0, 0], [-1, 0], [-1, -1], [0, 2], [-1, 2]],
  '3>0': [[0, 0], [-1, 0], [-1, -1], [0, 2], [-1, 2]],
  '0>3': [[0, 0], [1, 0], [1, 1], [0, -2], [1, -2]],
};

const I_KICKS: Record<string, Kick[]> = {
  '0>1': [[0, 0], [-2, 0], [1, 0], [-2, -1], [1, 2]],
  '1>0': [[0, 0], [2, 0], [-1, 0], [2, 1], [-1, -2]],
  '1>2': [[0, 0], [-1, 0], [2, 0], [-1, 2], [2, -1]],
  '2>1': [[0, 0], [1, 0], [-2, 0], [1, -2], [-2, 1]],
  '2>3': [[0, 0], [2, 0], [-1, 0], [2, 1], [-1, -2]],
  '3>2': [[0, 0], [-2, 0], [1, 0], [-2, -1], [1, 2]],
  '3>0': [[0, 0], [1, 0], [-2, 0], [1, -2], [-2, 1]],
  '0>3': [[0, 0], [-1, 0], [2, 0], [-1, 2], [2, -1]],
};

// tetr.io SRS+ 180 kicks (all pieces; I uses the same table there).
const KICKS_180: Record<string, Kick[]> = {
  '0>2': [[0, 0], [0, 1], [1, 1], [-1, 1], [1, 0], [-1, 0]],
  '2>0': [[0, 0], [0, -1], [-1, -1], [1, -1], [-1, 0], [1, 0]],
  '1>3': [[0, 0], [1, 0], [1, 2], [1, 1], [0, 2], [0, 1]],
  '3>1': [[0, 0], [-1, 0], [-1, 2], [-1, 1], [0, 2], [0, 1]],
};

export function kicksFor(type: PieceType, from: Rot, to: Rot): Kick[] {
  const key = `${from}>${to}`;
  if (type === 'O') return [[0, 0]];
  if ((from + 2) % 4 === to) return KICKS_180[key] ?? [[0, 0]];
  return (type === 'I' ? I_KICKS : JLSTZ_KICKS)[key] ?? [[0, 0]];
}

export interface RotateResult {
  x: number;
  y: number;
  rot: Rot;
  kickIndex: number;
}

/** Attempt rotation with kicks; returns the new position or null if blocked. */
export function tryRotate(
  board: Board,
  type: PieceType,
  rot: Rot,
  x: number,
  y: number,
  dir: 1 | -1 | 2,
): RotateResult | null {
  const to = ((rot + dir + 4) % 4) as Rot;
  const kicks = kicksFor(type, rot, to);
  for (let i = 0; i < kicks.length; i++) {
    const [dx, dy] = kicks[i];
    if (!board.collides(cellsAt(type, to, x + dx, y + dy))) {
      return { x: x + dx, y: y + dy, rot: to, kickIndex: i };
    }
  }
  return null;
}

File: rust/lst-solver/src/srs.rs
//! SRS kick tables with tetr.io-style 180 kicks. Port of src/core/srs.ts.
//! Offsets are (dx, dy) with y up, tried in order.

use crate::board::Board;
use crate::pieces::{cells_at, PieceType};

type Kick = (i32, i32);

// JLSTZ kicks keyed by (from, to)
fn jlstz_kicks(from: usize, to: usize) -> &'static [Kick] {
    match (from, to) {
        (0, 1) => &[(0, 0), (-1, 0), (-1, 1), (0, -2), (-1, -2)],
        (1, 0) => &[(0, 0), (1, 0), (1, -1), (0, 2), (1, 2)],
        (1, 2) => &[(0, 0), (1, 0), (1, -1), (0, 2), (1, 2)],
        (2, 1) => &[(0, 0), (-1, 0), (-1, 1), (0, -2), (-1, -2)],
        (2, 3) => &[(0, 0), (1, 0), (1, 1), (0, -2), (1, -2)],
        (3, 2) => &[(0, 0), (-1, 0), (-1, -1), (0, 2), (-1, 2)],
        (3, 0) => &[(0, 0), (-1, 0), (-1, -1), (0, 2), (-1, 2)],
        (0, 3) => &[(0, 0), (1, 0), (1, 1), (0, -2), (1, -2)],
        _ => &[(0, 0)],
    }
}

fn i_kicks(from: usize, to: usize) -> &'static [Kick] {
    match (from, to) {
        (0, 1) => &[(0, 0), (-2, 0), (1, 0), (-2, -1), (1, 2)],
        (1, 0) => &[(0, 0), (2, 0), (-1, 0), (2, 1), (-1, -2)],
        (1, 2) => &[(0, 0), (-1, 0), (2, 0), (-1, 2), (2, -1)],
        (2, 1) => &[(0, 0), (1, 0), (-2, 0), (1, -2), (-2, 1)],
        (2, 3) => &[(0, 0), (2, 0), (-1, 0), (2, 1), (-1, -2)],
        (3, 2) => &[(0, 0), (-2, 0), (1, 0), (-2, -1), (1, 2)],
        (3, 0) => &[(0, 0), (1, 0), (-2, 0), (1, -2), (-2, 1)],
        (0, 3) => &[(0, 0), (-1, 0), (2, 0), (-1, 2), (2, -1)],
        _ => &[(0, 0)],
    }
}

// tetr.io SRS+ 180 kicks (all pieces; I uses the same table there).
fn kicks_180(from: usize, to: usize) -> &'static [Kick] {
    match (from, to) {
        (0, 2) => &[(0, 0), (0, 1), (1, 1), (-1, 1), (1, 0), (-1, 0)],
        (2, 0) => &[(0, 0), (0, -1), (-1, -1), (1, -1), (-1, 0), (1, 0)],
        (1, 3) => &[(0, 0), (1, 0), (1, 2), (1, 1), (0, 2), (0, 1)],
        (3, 1) => &[(0, 0), (-1, 0), (-1, 2), (-1, 1), (0, 2), (0, 1)],
        _ => &[(0, 0)],
    }
}

pub fn kicks_for(t: PieceType, from: usize, to: usize) -> &'static [Kick] {
    if t == PieceType::O {
        return &[(0, 0)];
    }
    if (from + 2) % 4 == to {
        return kicks_180(from, to);
    }
    if t == PieceType::I {
        i_kicks(from, to)
    } else {
        jlstz_kicks(from, to)
    }
}

pub struct RotateResult {
    pub x: i32,
    pub y: i32,
    pub rot: usize,
    pub kick_index: i32,
}

/// Attempt rotation with kicks; returns the new position or None if blocked.
/// dir: 1 = CW, -1 (as 3) = CCW, 2 = 180.
pub fn try_rotate(
    board: &Board,
    t: PieceType,
    rot: usize,
    x: i32,
    y: i32,
    dir: i32,
) -> Option<RotateResult> {
    let to = (((rot as i32 + dir) % 4 + 4) % 4) as usize;
    let kicks = kicks_for(t, rot, to);
    for (i, &(dx, dy)) in kicks.iter().enumerate() {
        if !board.collides(&cells_at(t, to, x + dx, y + dy)) {
            return Some(RotateResult {
                x: x + dx,
                y: y + dy,
                rot: to,
                kick_index: i as i32,
            });
        }
    }
    None
}
